Nats Streaming Producer

jainnidhi703 opened the pull request:
**What**
Nats Streaming Async Producer (#1111)

**How**
Using the [Nats Java Connector](https://github.com/nats-io/java-nats-streaming) publishing message to Nats

**Usage**
Set `producer=nats` and other config related to nats: 
```
nats_url=<nats_url>
nats_cluster_id=<cluster_id>
nats_client_id=<client_id>
nats_subject=<subject>
```

Issue

osheroff commented on src/main/java/com/zendesk/maxwell/producer/MaxwellNatsProducer.java:
```diff
@@ -0,0 +1,88 @@
+		try {
+			this.connection = cf.createConnection();
+		} catch (IOException | InterruptedException e) {
+			logger.error("Exception during Nats Connection", e);
```
we'd probably want to throw a `RuntimeException` here.  There's no sense in carrying on with a broken connection.

osheroff commented on src/main/java/com/zendesk/maxwell/producer/MaxwellNatsProducer.java:
```diff
@@ -0,0 +1,88 @@
+
+	@Override
+	public void sendAsync(RowMap r, CallbackCompleter cc) throws Exception {
+		String msg = r.toJSON();
```
please use the call of `toJSON` that takes `outputConfig`

osheroff commented on src/main/java/com/zendesk/maxwell/producer/MaxwellNatsProducer.java:
```diff
@@ -0,0 +1,88 @@
+	@Override
+	public void sendAsync(RowMap r, CallbackCompleter cc) throws Exception {
+		String msg = r.toJSON();
+		NatsAckHandler handler = new NatsAckHandler(cc, r.getNextPosition(), msg, context);
```
you probably mean `getPosition` not `getNextPosition`

jainnidhi703 replied:
When I check the usage of getNextPosition in Producers such as MaxwellKinesisProducer, MaxwellKafkaProducer (classes extending AbstractAsyncProducer) all of them use getNextPosition. Redis producer is also using the same.

osheroff replied:
yes, they use `getNextPosition` to advance the binlog, but if you're just logging the position of the event, that position is `getPosition`....

jainnidhi703 commented:
@osheroff I didn't understand the use case of Stoppable task, although I implemented it-so when Nats disconnects taskState.stopped(), but when it reconnects the I wasn't able to change taskState to running again. Changed it to a boolean flag isNatsConnected instead.

osheroff commented on src/main/java/com/zendesk/maxwell/MaxwellConfig.java:
```diff
@@ -405,6 +422,15 @@ private void setup(OptionSet options, Properties properties) {
+		this.natsUrl = fetchOption("nats_url", options, properties, "nats://localhost:4222");
+		this.natsClusterId = fetchOption("nats_cluster_id", options, properties, "test-cluster");
+		this.natsClientId = fetchOption("nats_client_id", options, properties, "test-client");
+		this.natsSubject = fetchOption("nats_subject", options, properties, "test-subject");
```
sorry, just got back to this PR.  Putting a default here of `test-subject` seems a touch optimistic, I think probably better to have it required?  I guess a default of `maxwell` wouldn't be terrible either.

osheroff commented on src/main/java/com/zendesk/maxwell/MaxwellConfig.java:
```diff
@@ -405,6 +422,15 @@ private void setup(OptionSet options, Properties properties) {
+		this.natsClientId = fetchOption("nats_client_id", options, properties, "test-client");
+		this.natsSubject = fetchOption("nats_subject", options, properties, "test-subject");
+		this.natsPubAckTimeout = fetchLongOption("nats_pub_ack_timeout", options, properties, null);
+		this.natsMaxReconnects = Integer.parseInt(fetchOption("nats_max_reconnects", options, properties, "0"));
```
could just use `(Integer) fetchLongOption`...

osheroff commented on src/main/java/com/zendesk/maxwell/producer/MaxwellNatsProducer.java:
```diff
@@ -0,0 +1,229 @@
+				} else if (connectionListener.isNatsRunning()){
+					RowMap row = queue.take();
+					this.push(row);
+				}
```
this will be a very tight loop if nats is in a not-yet-connected state, we probably want to sleep for a few milliseconds if we're waiting for the connection to come online, no?

osheroff commented on src/main/java/com/zendesk/maxwell/producer/MaxwellNatsProducer.java:
```diff
@@ -0,0 +1,229 @@
+				LOGGER.error("Nats connection disconnected");
+				if (!context.getConfig().ignoreProducerError) {
+					this.context.terminate(new RuntimeException("Nats Disconnected"));
+					return;
```
should this be something that crashes maxwell?  Or can we wait for a reconnect?

jainnidhi703 replied:
if ignoreProducerError is false, then we should terminate right?

osheroff replied:
I'm just not sure what the `DISCONNECTED` message signifies: does it mean the client disconnected and will try to re-open the connection, or does it mean that it's dead and has given up?

jainnidhi703 replied:
DISCONNECTED callback comes when the nats connection was down.

jainnidhi703 commented:
@osheroff Please review the changes.

Thanks,
Nidhi

osheroff commented on src/main/java/com/zendesk/maxwell/producer/MaxwellNatsProducer.java:
```diff
@@ -0,0 +1,242 @@
+		try {
+			Thread.sleep(1000L);
+			while ( true ) {
+				// Irrespective of Nats is working or not, heartbeat should be pushed
```
comment in the wrong place now

osheroff commented on src/main/java/com/zendesk/maxwell/producer/MaxwellNatsProducer.java:
```diff
@@ -0,0 +1,242 @@
+			}
+		} catch ( Exception e ) {
+			LOGGER.info("Error: ", e);
+		}
```
can you look at `MaxwellKafkaProducer` for an example on how to handle errors, as well as how to exit gracefully (implement StoppableTask)

jainnidhi703 replied:
While using the StoppableTask, when Nats sent Reconnected event the task wasn't able to restart and push to Nats.

osheroff replied:
that doesn't make much sense to me.  StoppableTask really is just an interface that lets the rest of maxwell shutdown a producer/other task when it's time to exit.

osheroff commented:
are you running this in production anywhere?  I generally love code to be run in production before merging.

jainnidhi703 commented:
Not running it in production

osheroff commented:
running it in development, then?  anywhere?

jainnidhi703 commented:
Was hoping to get it merged and then deploy maxwell daemon with Nats feature

osheroff commented:
Unless it's trivial, I really really love to have code run somewhere before merged.  I can mark the producer as beta I guess, but don't you *also* want to know your code works?

osheroff commented:
nats done got implemented elsewhere
